### ISSUE TITLE
support multiple music_directory paths, including ones not necessarily present at start time

### ISSUE THREAD

**Issue by twojstaryzdomu** [MusicPlayerDaemon/MPD#1252]
## Feature request
MPD appears to support a single music_directory path. Enabling additional directories via multiple music_directory calls would be a convenient feature.

Failure to find the directory at start time should not be taken as assumption the path won't become available during the lifetime of the MPD process. 

The simplest use case for missing paths at start time is for use with removable media, which sit in mountpoints under /media/username, which may not be present if removable media isn't inserted. It would be up to the user to scan the media, so polling for changes (auto_update) need not apply to the scenario.

**Comment by MaxKellermann** [comment]
How would this work, at the protocol level?
Do you know that MPD can integrate with udisks2 to integrate removable media into its database?

**Comment by twojstaryzdomu** [comment]
Reading through the code, I've just stumbled upon the udisks2 plugin. Does it support any media inserted, by specifying udisks://? If the label of the media has to be known upfront, it may be difficult to use without access to the MPD configuration (other than a rescan option, which again may be automated by running a including udev rule whenever a user inserts or removes removable media).

At protocol level, supporting multiple paths, with non-existent ones at mpd start time, it may require two changes, if there isn't already code present to enable those:

1. MPD needs to virtualise all music_directory paths provided in a common structure. Failure to find a path should prompt for MPD to recorde for a subsequent rescan.
2. Any non-existent paths are assumed to be unpollable and excluded from auto_update.
3. At user requested rescan time, MPD attempts to walk all the paths including non-existing ones. Existing paths that become non-existent are flagged as non-existent and their music directory contents removed until a future rescan.

**Comment by MaxKellermann** [comment]
> Reading through the code, I've just stumbled upon the udisks2 plugin. Does it support any media inserted, by specifying `udisks://`, without the label or name?

No. You can obtain a list of available media using the `listneighbors` command, see https://mpd.readthedocs.io/en/latest/protocol.html#mounts-and-neighbors because udisks is both a storage and a neighbor plugin, see https://mpd.readthedocs.io/en/latest/plugins.html#neighbor-plugins

> At protocol level, supporting multiple paths, with non-existent ones at mpd start time, may require three changes, if there isn't already code present to enable those:

None of those 3 changes describe how this works at the protocol level, because none of this is about the protocol.

When you do "lsinfo", which music_directory will be displayed?

**Comment by twojstaryzdomu** [comment]
It wasn't clear it was MPD internal protocol that you meant. 

```
# egrep 'media|local' /etc/mpd.conf;
include_optional        "/etc/mpd.conf.d/media.conf"
include_optional        "/etc/mpd.conf.d/local.conf"
# cat /etc/mpd.conf.d/media.conf;
music_directory "/media/<username>"
# cat /etc/mpd.conf.d/local.conf
music_directory "/misc/local"
```

lsinfo displays whichever directory is configured last as music_directory. Any earlier directory does not appear to be looked at. If media directory (one possibly being a non-existent path) is provided earlier, it isn't ever examined, even at user requested rescan. Granted, if the removable media path is provided last, it is eventually examined at rescan, even if not present at MPD start time.

Still, only a single path is possible. True, protocol may require some constructs for adding/removing paths. To extend the present syntax, for adding & removing paths, the following config file parameters could work:

```
music_directory+ <path>
music_directory- <path>
```

That way, the compatibility with the earlier behaviour of music_directory, where the last parameter provided overrides earlier ones, would be maintained.

**Comment by muesli4** [comment]
I just tested the udisks neighbor plugin and it seems to work fine.  I guess the only issue related to this is the static nature of the current approach.  This could be solved if the clients themselves implement the neighbor and mounting functionality.  But this still puts a burden on the user to manually add it every time and the client.

One could write a separate MPD client to deal with neighbor events and then just mount them and do some action.  I guess this is the least invasive solution.  Otherwise, MPD itself could provide this _auto-mount_ functionality.  But I do not know how it works with the other neighbor plugins.

And then there is the issue about what should happen if you have songs in the queue or some playlist and unplug it (see https://github.com/MusicPlayerDaemon/MPD/issues/1142).

**Comment by twojstaryzdomu** [comment]
All of tgis

> I just tested the udisks neighbor plugin and it seems to work fine. I guess the only issue related to this is the static nature of the current approach. This could be solved if the clients themselves implement the neighbor and mounting functionality. But this still puts a burden on the user to manually add it every time and the client.
> 
> One could write a separate MPD client to deal with neighbor events and then just mount them and do some action (see blocker here #1270). I guess this is the least invasive solution. Otherwise, MPD itself could provide this _auto-mount_ functionality. But I do not know how it works with the other neighbor plugins.
> 
> And then there is the issue about what should happen if you have songs in the queue or some playlist and unplug it (see #1142).

Clients typically lag behind in functionality and tend to overlook features overly complex to implement or such that users won't ever look at.

While a udisks plugin certainly showcases functionality of MPD, it won't be used much unless you are in the position to perform media discovery, uuid look ups only only to mount a usb drive. Not many users will. Many (like myself) will depend on a set and forget configuration.

This request isn't about removable media anymore. It's up to the user what directories to set up for MPD. The existing removable media mounting mechanism in the OS distribution is perfectly fine for the job of mounting media. No plugins are strictly necessary to support adding/removing multiple media. Let KISS prevail here. Many unnecessary problems mentioned in your post would be avoided, had there been support for multiple directories in MPD.

**Comment by muesli4** [comment]
> Clients typically lag behind in functionality and tend to overlook features overly complex to implement or such that users won't ever look at. 
> [..]
> This request isn't about removable media anymore. It's up to the user what directories to set up for MPD. The requested feature would benefit any use case with music in multiple directories.

I would assume since udisks2 is also available as storage plugin you should be able to set it up in the configuration exactly as you intend to.  There is just a little bit of documentation lacking there.

**Comment by twojstaryzdomu** [comment]
> I would assume since udisks2 is also available as storage plugin you should be able to set it up in the configuration exactly as you intend to.

No, I'm not aware the udisks plugin works that way. You need to list the media manually with one plugin then mount them or input the device UUID upfront. Too much hassle if the goal is to access the media already automounted under `/media` by the OS. 

> There is just a little bit of documentation lacking there.

Indeed, the non-existent documentation is poor encouragement for the layman to try them out. I did and these plugins are impractical in an static configuration environment, with limited access by users.

Most clients support rescanning so users can perform that step easily themselves without raw access to the obscure interface that plugins currently use. The OS is doing the automounting just fine. All that is missing is support for multiple music_directory locations (with one of them being `/media` in my use case).

Given the multitide of options in MPD, partitions, multiple music outputs, etc., I was surprised NOT to find support for multiple music directories.

**Comment by maxmitti** [comment]
Correct me if I’m wrong, but can’t you just mount all dynamic mountpoints inside the configured music_directory, or alternatively somewhere else and add symlinks inside the music_directory pointing to the mountpoints?

The same works for multiple directories in general.

On changes you just trigger a rescan as you already suggested.

**Comment by bhurt1** [comment]
I have a no brainer solution, create a directory inside your music folder, I call it add, and just simply symlink all your music files there. I know its sounds too easy a solution when we could spend now 42 comments on how to solve an easily solvable problem. I'll do that while you rack your brains and can't get a working solution.

**Comment by muesli4** [comment]
> > It's the functionality that is already there. I guess some users may want this but you would have to ask @MaxKellermann.
> 
> Do not assume what is good for a developer is good for a user. Requiring mounts with its over-the-top syntax for more than 1 music source is bad design, not consistent with a single music_directory and as such not for the casual user. For developers orientating their clients around it is perfectly fine.

It seems to me you don't even know the protocol or what *mount* means in MPD.  As a first step you should read the manual.
 
> > MPD already has a virtual filesystem in the sense of a library. The root of which is the directory specified. Thus, changing this would break existing functionality.
> 
> How would mounting as per these examples exactly break existing functionality and which functionality precisely?

The functionality I was referring to is using the name of the mount point as the first path segment.  This is contradictory to how you can *mount* at the moment (at any place in the virtual file system of MPD).  And without that, you have to disambiguate, because files with the same name may exist in the same path.

> > It does not tell the user how they are accessible. And the mount points are not internal to MPD they are visible in the library
> 
> Explain what a user is not told following my proposal how to handle it. On the contrary, you get a full path to the music source first, followed by a relative path within the source. That serves hierarchical record and sufficiently unique.

The name of mounts are sometimes a bit cryptic.  Using them as first path segments may actually be confusing for users.  If the user does not care it is better to mount everything to the top level directory:
```
mount a
mount b
```
However, if the user cares, he can optionally specify where in the virtual file system of MPD he wants parts to be mounted:
```
mount a /wife/
mount b /mine/
mount c /mine/game-soundtracks/
```
With symbolic links you have undefined behavior with shadowing (or priority) and it is much harder to keep fresh if some of the directories change.  Also, imagine, your drive is not ready yet or you use a more exotic source, you can't do that with symlinks.  However, with this, everything is handled by MPD.

> > You could make the same argument about unix filesystems. I do not think it is bad design per se. You always need to think about what the requirements were when it was designed.
> 
> mpd isn't exactly a file system manager nor should it strive to become one. Intending to duplicate unix file systems functionality in mpd and requiring users to follow the same semantics (mountpoints) is over-engineering and ignoring the use case of a music player with a single hierarchy of tracks available across all devices.

Completely irrelevant.  It's there and it is a useful feature.  Make yourself familiar with the feature before discussing about it.

> Some common config file parameters, like `password` or `metadata_to_use`, take reasonable syntax to perform the necessary work behind the scenes. Other parameters have syntax that varies with type, like `audio_output`, and as such it's reasonable to have its syntax implemented as a hash/map. For multiple music sources in `music_directory`, the mountpoint can be easily derived from the mount source, so it's more like the former. Besides, complex syntax invites user issues.

Except that it is not complex and this YAML-style configuration is used in many other parts of the MPD config already.

**Comment by bhurt1** [comment]
Exactly what I said, you just overcomplicated something that needn't be. Let's be clear, I understand all of this stuff but I live by the KISS metric. Unless you like to be the last mnemonic of that phrase. And, I don't like sym links as a solid solution, but here it works for me. Exotic, you must be pluggin and rewritin at speeds that is stupendous, again keep it simple, and don't get stupid, with organization it can be handled handily. I stand by my long standing mnemonic, KISS.

**Comment by twojstaryzdomu** [comment]
@bhurt1 cut the simpleton sarcasm. Links may indeed be simpler to undersand but don't cover all cases. You don't always have the luxury of defining those, such as when the source isn't a local path (say you mix nfs based sources) or no file system access, except for the edit ability of the music_directory field. Technically there is autofs but it's terribly unreliable (though improving) and users would need root to tinker with it anyway. The mount command is bad design - something an overly eager developer would expose to the world when it is targetted at developers designing their clients to talk to mpd. The labels mentioned above are unnecessarily duplicating a  POSIX file system paradigm in a music client. And writing xmls or jsons is an exercise in troubleshooting bad syntax for errors and typos. Most users don't wish to read documentation to accomplish an ordinary task. MPD is otherwise a player that does its job but due to such cruft it is still rough around the edges.

Symbolic links in a single directory enable a facility externally but have som0e limitations (you don't know what sources a user will want to expose to mpd). Suppose your link points to a protocol name like a music streaming service or nfs://, does it work for you?
Contrast that with updating the syntax of a single command to include mulitiple semicolon separated items, of any type possible. And speaking of KISS, you mistake your no brainer with a bodge which is what links are. All configuration should ideally remain in one place. Not part configuration file, part file system placement or internal command dressed as a unix style command, without much need for it.

Here's KISS at it's simplest:

music_directory="/music;nfs://server:/music2;streaming_servicename://credentials ... 

Except to pull that off, you'd need to stop thinking as a developer or a unix file system enthusiast and consider how other established linux software defines it configuration.
Do we have xmls or jsons in systemd configuration? No, because it's one thing they did right. Jsons are a programmer's way to indicate they want to force their users to study their painstakingly written documentation.

**Comment by bhurt1** [comment]
I think you answered your pillory of my answer, and I mean that, although probably more correctly the pillory of this problem. The problem doesn't exist, it's been livin rent free in your head for yrs now. somehow you can't accept that it can be solved simply., And if it can't, cuz there are limitations, build your own or move on, that's what linus did, and here we are, you and me, you are so wrapped in bs you can't think straight.

**Comment by twojstaryzdomu** [comment]
> somehow you can't accept that it can be solved simply.

As stated, it doesn't cover all cases, anything that's not within the local file system won't work. I'd like links to work like you propose but it just doesn't work that way (yet).

>  And if it can't, cuz there are limitations, build your own or move on

Or open an issue. If everyone just shrugged their shoulders like you propose, many sensible ideas wouldn't make it. The fact a poor man's work-around of sorts exists doesn't make it any less of a problem.

> it's been livin rent free in your head for yrs now. 
> you are so wrapped in bs you can't think straight.

On the one hand you raise a fair point but on the other you seem a plain troll. This just isn't the place for sorting out personality issues.

**Comment by bhurt1** [comment]
you are a pure degenerate, you call features of linux that are by far superior to win, a poor mans work around, and you say I am sorting out personality issues ? Its obvious you are projecting by confession. I understand I covered more ground in fewer words than you can say  in **** ****, a few explitives. so deal with it. As is painfully obvious and your need to try and call ppl stupid or simpleton, as you said, is again confession by projection, a common liberal tactic. So I ask are you going to be able to work for the next month after I have talked this way to you, like a true liberal twitter lord.

**Comment by twojstaryzdomu** [comment]
> you are a pure degenerate
How old are you? Your manner of talk is clearly intended to abuse as such it has no place here. Such statements appear to come from someone in need of an anger management class not a person with anything of value to add.

 > **** ****
Self censorship? No consistency in your sworn use of expletives.

> call ppl stupid or simpleton
Not stupid and not people, but your sarcastic talk. Read what I wrote again.

**Comment by MaxKellermann** [comment]
@twojstaryzdomu don't waste your time with pricks like that. I have blocked them now.

**Comment by twojstaryzdomu** [comment]
> @twojstaryzdomu don't waste your time with pricks like that. I have blocked them now.

No needed to block them. Albeit arrogantly, he did make a point about symbolic links being a possible work around to the `music_directory` restriction. I opened #1535 to request resolution of symlinked paths to plugin-based sources.

**Comment by twojstaryzdomu** [comment]
[Alternative proposal](https://github.com/MusicPlayerDaemon/MPD/issues/1537) to enable multiple music sources non-interactively.

**Comment by stardiviner** [comment]
What about a map structure value for multiple music directories?
```conf
music_directory {
  "path 1"
  "path 2"
}
```

Create a symbolic link to music directory is a temporary solution. But not ideal for all solution.
For example, SMB mounted folder on NAS1, and NFS mounted folder on NAS2.
